### PR TITLE
Changed dict key-matching regex to capture any valid dict key

### LINKED ISSUE #917: Tab completion for dict key causes crash
This issue is not reproducible with several different `dict`s that I've tried, but it can be reproduced with the `dict` below.

```
my_dict = {
'Circle': {'Ellipse',
           'Shape',
           'ShapeDriver'},

'Ellipse': {'Shape'},

'Rectangle': {'Shape'},

'Shape': {'Circle'},

'ShapeDriver': {'Circle',
                'Ellipse',
                'Shape'},

'Square': {'Rectangle'}
}
```

Steps to recreate:
1. create new `dict`
2. type `my_dict[`
3. press <tab> key to use tab-completion
4. crash

```
~ % bpython
bpython version 0.21 on top of Python 3.8.2 /Library/Developer/CommandLineTools/usr/bin/python3
>>> my_dict = {
... 'a/Circle': {'Ellipse',
...            'Shape',
...            'ShapeDriver'},
... 
... 'a/Ellipse': {'Shape'},
... 
... 'a/Rectangle': {'Shape'},
... 
... 'a/Shape': {'Circle'},
... 
... 'a/ShapeDriver': {'Circle',
...                 'Ellipse',
...                 'Shape'},
... 
... 'a/Square': {'Rectangle'}
... }
>>> my_dict['b/banana']
Traceback (most recent call last):
  File "/Users/arian/Library/Python/3.8/bin/bpython", line 8, in <module>
    sys.exit(main())
  File "/Users/arian/Library/Python/3.8/lib/python/site-packages/bpython/curtsies.py", line 201, in main
    exit_value = repl.mainloop(True, paste)
  File "/Users/arian/Library/Python/3.8/lib/python/site-packages/bpython/curtsies.py", line 121, in mainloop
    self.process_event_and_paint(e)
  File "/Users/arian/Library/Python/3.8/lib/python/site-packages/bpython/curtsies.py", line 85, in process_event_and_paint
    self.process_event(e)
  File "/Users/arian/Library/Python/3.8/lib/python/site-packages/bpython/curtsiesfrontend/repl.py", line 615, in process_event
    return self.process_key_event(e)
  File "/Users/arian/Library/Python/3.8/lib/python/site-packages/bpython/curtsiesfrontend/repl.py", line 757, in process_key_event
    self.on_tab()
  File "/Users/arian/Library/Python/3.8/lib/python/site-packages/bpython/curtsiesfrontend/repl.py", line 891, in on_tab
    cursor_and_line = self.matches_iter.substitute_cseq()
  File "/Users/arian/Library/Python/3.8/lib/python/site-packages/bpython/repl.py", line 297, in substitute_cseq
    self.update(
  File "/Users/arian/Library/Python/3.8/lib/python/site-packages/bpython/repl.py", line 318, in update
    self.start, self.end, self.current_word = self.completer.locate(
TypeError: cannot unpack non-iterable NoneType object
```

Image below so you can see the colors as well:
image

### PR DESCRIPTION
**Pull request by arian-deimling**
Fixes #917

### CODE REVIEW

**Comment by thomasballinger** [comment]
I tried this out and found that this more general pattern meant the key matching too much of the line, so tab completion of the second dictionary on a line no longer works. I think `.*` is too broad a pattern, could we restrict it further? Maybe we could allow anything if it started with a `'` or `"` and once that quote is matched, not allow more?



It would be nice to add some tests, both of the new functionality you're adding (backslashes in strings work!) and this regression (second dictionary on a line should still work) like https://github.com/bpython/bpython/blob/b66a29fde838f98a39c08ce1b94f6cc8b8c3e86d/bpython/test/test_line_properties.py#L171-L184
These tests use a DSL to describe the expected match when calling one of these functions:
https://github.com/bpython/bpython/blob/b66a29fde838f98a39c08ce1b94f6cc8b8c3e86d/bpython/test/test_line_properties.py#L77-L86

so a test for this regression might look like
```
    self.assertAccess("(asdf['abc'], asdf[<'ab>|") 
```

Thanks for looking into this!

**Comment by thomasballinger** [comment]
@arian-deimling could you add tests to keep track of the different cases we've talked about?

This pattern doesn't match
```
>>> d = {'asdf': 123, 'a\'b"\\c': 234}
>>> d['a\'
```
but that didn't work before either, so I think your code is a winner!

You can run the tests by running `pytest` in the repo.

**Comment by arian-deimling** [comment]
@thomasballinger If you don't mind, I'd like like to work on this a little bit more. It's more nuanced than I originally thought, but I think with a few more minor modifications, we could have the tab completion working for most different types of potential dict keys (a.k.a. hashable types - https://stackoverflow.com/questions/14535730/what-does-hashable-mean-in-python/44880799#44880799). I agree that there will probably have to be a few edge cases that we cannot handle.

And to answer your first question, yes, I will add some tests!

**Comment by arian-deimling** [comment]
@thomasballinger Ok - ready for you to have another look with these two commits. This newest version should be able to tab-complete all `str`, `bytes`, and `int` keys, `float` keys besides special values (i.e. nan, inf, -inf), and `tuple` keys for tuples containing any combination of the aforementioned key possibilities.

Also, that testing stuff is pretty neat and straightforward. The changes pass all tests as well as 9 new ones. 

This may be off topic, but I think this test case is irrelevant:
https://github.com/bpython/bpython/blob/b66a29fde838f98a39c08ce1b94f6cc8b8c3e86d/bpython/test/test_line_properties.py#L176
because in the case where you have typed `some_dict[abc`, the program is not trying to match `abc` as a key, but rather as the beginning of a variable or function name.

**Comment by thomasballinger** [comment]
> This may be off topic, but I think this test case is irrelevant because in the case where you have typed some_dict[abc, the program is not trying to match abc as a key, but rather as the beginning of a variable or function name.

I agree that that code path isn't useful when running bpython, and I am ambivalent about inverting the test and changing the behavior. I wrote out the following to think through this.

---

If we implemented fuzzy completion for dict keys (which we do have implemented in `AttrCompletion` and `GlobalCompletion`) then it would useful to match this, assuming we still want DictKeyCompletion to take precedence over expression completion in that situation. I don't plan to do that though, and I'm not sure we'd want to make it so fuzzy that quotes will be added for you. If the completer were every changed to return matches that included global variables that referred to dictionary keys we would need it... but no plans for that. If we decided to show all matches instead of filtering on prefix we'd want this, but that doesn't sound very useful.

The only reason to leave this test case is that always matching the item in the dict key position makes it easier to describe what `DictKeyCompletion.locate()` does since its more in the spirit of what locate() is supposed to do according to the `BaseCompletion` class:

https://github.com/bpython/bpython/blob/b66a29fde838f98a39c08ce1b94f6cc8b8c3e86d/bpython/autocomplete.py#L225-L226

A reason to change the behavior and explicitly not match anything other than hashable literals is that it would be a more efficient solution in some cases than [what I implemented](https://github.com/bpython/bpython/commit/58f3edfb26875450bbd71b909cb003f6b942bdab) for deciding which completer to use: if `DictKeyCompletion.locate()` returned None, we could bail out of DictKeyCompletion.matches() before needing to evaluate the dictionary expression, get keys, and see if the expression is a prefix of any keys (and it never is, since the only candidate keys are literals). However if we want to bail out early we could just check manually in `DictKeyCompletion.matches` instead of changing what `locate()` means.

https://github.com/bpython/bpython/blob/b66a29fde838f98a39c08ce1b94f6cc8b8c3e86d/bpython/autocomplete.py#L653-L668

For now, starting to type an identifier in the dict key position means we will not be able to complete a dictionary key. So returning None for something that could not possibly be a prefix of a valid key literal does make sense: there is a dict key, but it's not a completable one and this code is all about completion. So I'd be ok with this change, but I think it's ok as-is too.

File: bpython/line.py [hunk 1]
```diff
@@ -34,7 +34,33 @@ def current_word(cursor_offset: int, line: str) -> Optional[LinePart]:
...
+    f'{_match_str_bytes_int_float}' \
+    r""", )*(?:""" \
+    f'{_match_str_bytes_int_float}' \
+    r"""\,|""" \
+    f'{_match_str_bytes_int_float}' \
+    r""")?\)?"""
+
+_current_dict_key_re_base = r"""[\w_][\w0-9._]*\["""
+_current_dict_key_re = LazyReCompile(
+    f'{_current_dict_key_re_base}('
```
Comment (thomasballinger) [inline bpython/line.py:56]: the "_base" part of _current_dict_key_re_base confuses me a little, _base because it's the beginning of the pattern? because the dict is the base of the expression?

**Review by thomasballinger — COMMENTED** [review]
Thanks for taking the time to get this right, this looks much more precise than we had it. I learned a bunch from reading it.

Does matching more types like this expand what works for substitutions on tab autocomplete? I'd like to understand the behavior change to figure out what the more complex code is buying us.

File: bpython/line.py [hunk 2]
```diff
@@ -34,7 +34,33 @@ def current_word(cursor_offset: int, line: str) -> Optional[LinePart]:
     return LinePart(start, end, word)
 
 
-_current_dict_key_re = LazyReCompile(r"""[\w_][\w0-9._]*\[([\w0-9._(), '"]*)""")
+# pieces of regex to match repr() of several hashable built-in types
+_match_identifiers_and_functions = r"""[a-zA-Z_][a-zA-Z0-9_.(),\[\] ]*"""
+_match_single_quote_str_bytes = \
+    r"""b?'(?:\\(?:\\\\)*['"nabfrtvxuU]|[^'\\])*(?:\\\\)*\\?'?"""
+_match_double_quote_str_bytes = r"""b?"[^"]*"?"""
```
[the remark this answers is not in the document]
Comment (arian-deimling) [reply bpython/line.py:41]: Yea, based on some research and just testing different strings, repr() will only surround a string in double-quotes if it contains at least one single quote char (`'`) and exactly zero double quote chars (`"`).

File: bpython/line.py (hunk 1 above)
Comment (arian-deimling) [reply bpython/line.py:56]: Yea, I could use some direction on the name if you don't like that, but basically, `base` is the part that I did not change which is everything before the `[` character after which 'key matching mode' is entered.

File: bpython/line.py [hunk 3]
```diff
@@ -34,7 +34,33 @@ def current_word(cursor_offset: int, line: str) -> Optional[LinePart]:
     return LinePart(start, end, word)
 
 
-_current_dict_key_re = LazyReCompile(r"""[\w_][\w0-9._]*\[([\w0-9._(), '"]*)""")
+# pieces of regex to match repr() of several hashable built-in types
+_match_identifiers_and_functions = r"""[a-zA-Z_][a-zA-Z0-9_.(),\[\] ]*"""
+_match_single_quote_str_bytes = \
+    r"""b?'(?:\\(?:\\\\)*['"nabfrtvxuU]|[^'\\])*(?:\\\\)*\\?'?"""
```
[the remark this answers is not in the document]
Comment (arian-deimling) [reply bpython/line.py:40]: Agreed. I will take care of that.

File: bpython/line.py [hunk 4]
```diff
@@ -34,7 +34,33 @@ def current_word(cursor_offset: int, line: str) -> Optional[LinePart]:
     return LinePart(start, end, word)
 
 
-_current_dict_key_re = LazyReCompile(r"""[\w_][\w0-9._]*\[([\w0-9._(), '"]*)""")
+# pieces of regex to match repr() of several hashable built-in types
```
[the remark this answers is not in the document]
Comment (arian-deimling) [reply bpython/line.py:37]: Good idea! I'll update it to include that.. honestly, it's even a little bit difficult for me to read even though I only wrote it a day ago.

**Comment by arian-deimling** [comment]
> Thanks for taking the time to get this right, this looks much more precise than we had it. I learned a bunch from reading it.
> 
> Does matching more types like this expand what works for substitutions on tab autocomplete? I'd like to understand the behavior change to figure out what the more complex code is buying us.

Yes, now you can tab-complete any of the types that I added a regex for, and the way I tested for this was to make a `dict` with those kinds of keys and then type `my_dict[` and then make sure that for every new character I typed in, the completion options were still displayed to the user.

Tuples are a special case because you cannot tab-complete them unless you type `my_dict[|]` with the pipe character representing the cursor location. I'm not sure exactly what the cause is, but it matches the behavior prior to my changes, so I think it's okay. One thing I noticed in these cases where you type the closing brace `[` is that when you use the tab completion to finish your key, bpython adds a closing brace `]`, so you are left with two `]`. We may want to open another issue for that; however, this behavior is also consistent with the main branch on the bpython repo.

File: bpython/line.py (hunk 1 above)
Comment (thomasballinger) [reply bpython/line.py:56]: maybe `_current_dict_key_dict_re`, or maybe since all of these substrings are not indented to be used separately, a different name like `_current_dict_key_dict_part` — OH or to match your current convention of `_match_*` maybe just `_dict` or `_dict_before_match` or something. 🤞 there'll be a beautiful was to do this with re.VERBOSE, but if not just changing to something shorter is fine. I just don't want to force readers to know the history of the code in order to understand the variable name.

Comment (arian-deimling) [reply bpython/line.py]: Renamed to `_match_dict_before_key`.

**Comment by arian-deimling** [comment]
@thomasballinger Okay.. so I think I lost sight of what issue I was trying to solve but I think this final commit is the one!

This comment is what I'm basing this last change on:
>The only reason to leave this test case is that always matching the item in the dict key position makes it easier to describe what DictKeyCompletion.locate() does since its more in the spirit of what locate() is supposed to do according to the BaseCompletion class

So the original issue was that we weren't matching everything that can possibly be in the key position, and the primary issue is strings because they can contain the `]` character which is the natural way to detect the end of the dict key. To fix this I realized, we should first capture any string in the key position, and outside of strings, we should capture anything but the `]` character. If we capture any number or combination of these using something like:
` (  (?: regex_to_match_any_string | regex_to_match_anything_but_close_square_bracket )*  )`
then we will capture the entire key. This has the added benefit of allowing text completion of for a multitude of other possible dict keys, such as `range`s, `complex` numbers among others that my older version did not capture.

There is one regression here, but I think it is worth the additional functionality. Current in `bpython`, if you test
`abcd[<'>|]` the test will pass, but with my new version the version of the test that would pass is `abcd[<'|]>`. This is because the `'` initiates string pattern matching and therefore the `]` is considered part of the string that is being typed in. This means that you cannot tab-complete a string if you've typed `my_dict[|]` and then start typing the string at the cursor position; however, this regression is offset by the fact that you can tab-complete a string that contains the `]` character.

Let me know what you think, but I think the solution to the regression is by making a change in the `locate` or `match` function down the line.

**Comment by arian-deimling** [comment]
@thomasballinger Can you re-run the test? I fixed the code to pass the black linter test.

**Comment by arian-deimling** [comment]
I didn’t mean to do that. Just clicked on the button to see what is does. Sorry

**Comment by thomasballinger** [comment]
No problem! I want Sebastian to take a look, and he's busy today so it'll be another day at least FYI.

File: bpython/line.py [hunk 5]
```diff
@@ -6,6 +6,7 @@
 
 from itertools import chain
 from typing import Optional, NamedTuple
+import re
```
Comment (sebastinas) [inline bpython/line.py:9]: Nitpick: we usually have `import` first, then `from ... import ...`.

File: bpython/test/test_line_properties.py [hunk 6]
```diff
@@ -178,10 +178,19 @@ def test_simple(self):
...
         # TODO self.assertAccess('d[d[<12|>')
         self.assertAccess("d[<'a>|")
+        self.assertAccess("object.dict['a'bcd'], object.dict[<'abc>|")
+        self.assertAccess(r"object.dict[<'a\'\\\"\n\\'>|")
+        self.assertAccess("object.dict[<\"abc'>|")
+        self.assertAccess("object.dict[<(1, 'apple', 2.134>|]")
+        self.assertAccess("object.dict[<(1, 'apple', 2.134)>|]")
+        self.assertAccess("object.dict[<-1000>|")
+        self.assertAccess("object.dict[<-0.23948>|")
+        self.assertAccess("object.dict[<'\U0001ffff>|")
```
Comment (sebastinas) [inline bpython/test/test_line_properties.py:193]: I think it would make sense to also add some tests for string keys containing `[` and `]` .

Comment (arian-deimling) [reply bpython/test/test_line_properties.py:194]: I misinterpreted this and added a test where the user is typing a string dict key inside `[` and `]` and found an issue so I fixed that by modifying the regex. Also added some tests for strings that contain `[` and `]`.

File: bpython/line.py (hunk 5 above)
Comment (arian-deimling) [reply bpython/line.py]: Done.

**Comment by arian-deimling** [comment]
@sebastinas I think it's good to go now.

**Comment by arian-deimling** [comment]
Turned off `black` globally because I don't like double quotes and forgot to reenable it for this @sebastinas. Good to go now. Sorry

**Comment by thomasballinger** [comment]
Thanks for the summary! Looks good to me.

> This means that you cannot tab-complete a string if you've typed my_dict[|] and then start typing the string at the cursor position; however, this regression is offset by the fact that you can tab-complete a string that contains the ] character.

That is unfortunate, I bet it's more common to autocomplete inside [] than to want to complete keys with ] in them; but it's more correct and if we want to fix this we can do it explicitly with a workaround instead of an accident of the way the code works. But there are other similar edge cases here, it would be neat if e.g. `my_dict["abc|"]` completed the key `"abcdef"` but prefix matching isn't going to do that. We can fix those sometime separately.
